### ISSUE TITLE
Long games 

### ISSUE THREAD

**Issue by sudosandwich** [gcp/leela-zero#1608]
I think this is a problem: 636 moves played in this training game. While playing to the end occasionally is of course important, at more than 1.5 times the available points to play, maybe we can decide by MC only. 600 moves for this game means we could have played 3 other games with 200 moves each. Is that assumption wrong? Any input appreciated.

Game:

{
    "cmd": "selfplay",
    "hash": "e1d466aa129ef0d4b52288d1051570396c69cea50a3cd28b45ae6c524119e22c",
    "hash_gzip_hash": "23c29bf777e446b5c3fb0e6e7fa4d53f15b99cc0c25798b70b57877b55bf1638",
    "minimum_autogtp_version": "16",
    "minimum_leelaz_version": "0.15",
    "options": {
        "noise": "true",
        "playouts": "0",
        "randomcnt": "999",
        "resignation_percent": "0",
        "visits": "3201"
    },
    "options_hash": "7aaaeb",
    "random_seed": "4447983281341852077",
    "required_client_version": "16"
}

Got new job: selfplay
net: e1d466aa129ef0d4b52288d1051570396c69cea50a3cd28b45ae6c524119e22c.
Engine has started.
Infinite thinking time set.
1 (B D3) 2 (W R16) 3 (B P16) 4 (W C16) 5 (B Q4) 6 (W P17) 7 (B O17) 8 (W Q17) 9 (B O16) 10 (W R14) 11 (B E16) 12 (W G17) 13 (B C17) 14 (W D16) 15 (B D17) 16 (W E17) 17 (B B16) 18 (W E15) 19 (B F16) 20 (W F17) 21 (B B15) 22 (W G16) 23 (B F15) 24 (W G14) 25 (B F14) 26 (W F13) 27 (B E14) 28 (W E13) 29 (B K17) 

30 (W G15) 31 (B D14) 32 (W C5) 33 (B D5) 34 (W D6) 35 (B C4) 36 (W E5) 37 (B D4) 38 (W D13) 39 (B C14) 40 (W B5) 41 (B E6) 42 (W D7) 43 (B E7) 44 (W F5) 45 (B D8) 46 (W C8) 47 (B E8) 48 (W F3) 49 (B C9) 50 (W B7) 51 (B B8) 52 (W C7) 53 (B F2) 54 (W G2) 55 (B B10) 56 (W A8) 57 (B E2) 58 (W B9) 59 (B G3) 60 (W G7) 61 (B H3) 62 (W M17) 63 (B M16) 64 (W M18) 65 (B Q16) 66 (W R17) 67 (B P13) 68 (W L16) 69 (B M15) 70 (W Q13) 71 (B L17) 72 (W O18) 73 (B K14) 74 (W P14) 75 (B G9) 76 (W J14) 77 (B K13) 78 (W F9) 79 (B F10) 80 (W J4) 81 (B J3) 82 (W K15) 83 (B J15) 84 (W J16) 85 (B K16) 86 (W H15) 87 (B L15) 88 (W K4) 89 (B R6) 90 (W O13) 91 (B G8) 92 (W K3) 93 (B N3) 94 (W G4) 95 (B K8) 96 (W N5) 97 (B M4) 98 (W H4) 99 (B H2) 100 (W M5) 101 (B P3) 102 (W J8) 103 (B J9) 104 (W J7) 105 (B K10) 106 (W R7) 107 (B Q7) 108 (W R8) 109 (B S6) 110 (W Q8) 111 (B K7) 112 (W L9) 113 (B K9) 114 (W J13) 115 (B L5) 116 (W K12) 117 (B L12) 118 (W L13) 119 (B K5) 120 (W L11) 121 (B P7) 122 (W P8) 123 (B O8) 124 (W O9) 125 (B N8) 126 (W N9) 127 (B M8) 128 (W B3) 129 (B M10) 130 (W C2) 131 (B K2) 132 (W L2) 133 (B G1) 134 (W K1) 135 (B J2) 136 (W M12) 137 (B N11) 138 (W M9) 139 (B L10) 140 (W O11) 141 (B T9) 142 (W S9) 143 (B T8) 144 (W T10) 145 (B S7) 146 (W S8) 147 (B T7) 148 (W S10) 149 (B C11) 150 (W E18) 151 (B N12) 152 (W N13) 153 (B G12) 154 (W G13) 155 (B D18) 156 (W D2) 157 (B E3) 158 (W E10) 159 (B E9) 160 (W D19) 161 (B C19) 162 (W E19) 163 (B B17) 164 (W D10) 165 (B M14) 

166 (W J11) 167 (B F11) 168 (W E11) 169 (B H12) 170 (W J12) 171 (B O12) 172 (W P12) 173 (B H11) 174 (W B12) 175 (B C12) 176 (W B11) 177 (B B4) 178 (W A4) 179 (B C3) 180 (W B2) 181 (B A3) 182 (W A2) 183 (B C13) 184 (W C10) 185 (B J15) 186 (W N17) 187 (B L18) 188 (W J18) 189 (B J17) 190 (W H17) 191 (B N16) 192 (W O15) 193 (B N18) 194 (W N19) 195 (B H16) 196 (W L14) 197 (B H18) 198 (W H19) 199 (B O10) 200 (W N10) 201 (B L19) 202 (W N18) 203 (B T16) 204 (W M11) 205 (B N15) 206 (W O6) 207 (B K15) 208 (W L3) 209 (B R2) 210 (W O7) 211 (B P15) 212 (W P10) 213 (B O12) 214 (W B18) 215 (B M19) 216 (W P18) 217 (B O14) 218 (W N14) 219 (B C18) 220 (W Q6) 221 (B K18) 222 (W J19) 223 (B T17) 224 (W Q2) 225 (B Q5) 226 (W P6) 227 (B K19) 228 (W G18) 229 (B N11) 230 (W N12) 231 (B Q15) 232 (W J16) 233 (B S11) 234 (W H6) 235 (B D12) 236 (W H8) 237 (B J6) 238 (W M13) 239 (B J1) 240 (W T5) 241 (B T15) 242 (W S13) 243 (B Q14) 244 (W P13) 245 (B A14) 246 (W S16) 247 (B O4) 248 (W B13) 249 (B S5) 250 (W M7) 251 (B O2) 252 (W L8) 253 (B N2) 254 (W R13) 255 (B F18) 256 (W B6) 257 (B P9) 258 (W L6) 259 (B L7) 260 (W B14) 261 (B R4) 262 (W D9) 263 (B M1) 264 (W H10) 265 (B H9) 266 (W J10) 267 (B T3) 268 (W G10) 269 (B T12) 270 (W F12) 271 (B F8) 272 (W R15) 273 (B H5) 274 (W A13) 275 (B A10) 276 (W O15) 277 (B G19) 278 (W L16) 279 (B A11) 280 (W A15) 281 (B A19) 282 (W G11) 283 (B Q3) 284 (W S18) 285 (B K11) 286 (W L12) 287 (B B19) 288 (W T2) 289 (B Q14) 290 (W Q19) 291 (B L1) 292 (W H7) 

293 (B T4) 294 (W M19) 295 (B P5) 296 (W F6) 297 (B Q11) 298 (W Q10) 299 (B L4) 300 (W K13) 301 (B T14) 302 (W R3) 303 (B Q18) 304 (W N4) 305 (B A6) 306 (W C6) 307 (B M3) 308 (W F9) 309 (B D11) 310 (W M14) 311 (B R18) 312 (W A3) 313 (B R1) 314 (W B8) 315 (B S15) 316 (W N11) 317 (B G5) 318 (W K16) 319 (B G2) 320 (W T13) 321 (B F4) 322 (W C9) 323 (B R5) 324 (W A17) 325 (B Q15) 326 (W F10) 327 (B D1) 328 (W N15) 329 (B L19) 330 (W K19) 331 (B J17) 332 (W K17) 333 (B A7) 334 (W Q12) 335 (B T19) 336 (W A9) 337 (B S17) 338 (W K6) 339 (B L17) 340 (W F19) 341 (B R19) 342 (W A16) 343 (B S4) 344 (W H16) 345 (B S12) 346 (W E12) 347 (B S1) 348 (W J5) 349 (B E4) 350 (W T18) 351 (B R12) 352 (W E1) 353 (B M2) 354 (W G6) 355 (B Q16) 356 (W pass) 357 (B M16) 358 (W N1) 359 (B P4) 360 (W O12) 361 (B N6) 362 (W K18) 363 (B M6) 364 (W S19) 365 (B P19) 366 (W L18) 367 (B N16) 368 (W L15) 369 (B O5) 370 (W O19) 371 (B Q9) 372 (W S3) 373 (B N5) 374 (W F7) 375 (B Q7) 376 (W K15) 377 (B N7) 378 (W K7) 379 (B B1) 380 (W A18) 381 (B T6) 382 (W P16) 383 (B E8) 384 (W C15) 385 (B G8) 386 (W P1) 387 (B T5) 388 (W P19) 389 (B T1) 390 (W T11) 391 (B O16) 392 (W C19) 393 (B Q18) 394 (W K11) 395 (B P7) 396 (W R11) 397 (B K9) 398 (W D17) 399 (B B16) 400 (W C18) 401 (B O7) 402 (W H5) 403 (B K8) 404 (W K14) 405 (B R18) 406 (W A5) 407 (B M10) 408 (W R9) 409 (B H14) 410 (W C1) 411 (B J9) 412 (W E7) 413 (B O6) 414 (W O1) 415 (B A6) 416 (W B15) 417 (B S12) 418 (W P2) 419 (B E9) 420 (W L10) 421 (B P6) 

422 (W H18) 423 (B Q6) 424 (W D1) 425 (B H9) 426 (W J17) 427 (B K10) 428 (W B17) 429 (B F8) 430 (W G9) 431 (B S11) 432 (W G5) 433 (B H1) 434 (W O14) 435 (B K1) 436 (W L17) 437 (B A19) 438 (W F11) 439 (B M5) 440 (W A14) 441 (B K9) 442 (W D18) 443 (B O17) 444 (W D8) 445 (B T12) 446 (W J9) 447 (B L7) 448 (W M15) 449 (B O17) 450 (W R12) 451 (B P9) 452 (W N16) 453 (B D15) 454 (W S12) 455 (B G8) 456 (W L19) 457 (B M7) 458 (W T19) 459 (B F1) 460 (W pass) 461 (B F8) 462 (W M10) 463 (B S2) 464 (W R19) 465 (B K8) 466 (W R10) 467 (B E9) 468 (W S14) 469 (B pass) 470 (W A12) 471 (B pass) 472 (W A11) 473 (B B10) 474 (W Q18) 475 (B R3) 476 (W E6) 477 (B T2) 478 (W T14) 479 (B F3) 480 (W A1) 481 (B T17) 482 (W Q9) 483 (B pass) 484 (W J6) 485 (B S17) 486 (W M16) 487 (B S3) 488 (W E8) 489 (B Q1) 490 (W O10) 491 (B T15) 492 (W K10) 493 (B O1) 494 (W H13) 495 (B K8) 496 (W F8) 497 (B P2) 498 (W H11) 499 (B N4) 500 (W A10) 501 (B P1) 502 (W pass) 503 (B T16) 504 (W B16) 505 (B G12) 506 (W H9) 507 (B O3) 508 (W H12) 509 (B pass) 510 (W J15) 511 (B pass) 512 (W T12) 513 (B Q2) 514 (W N1) 515 (B L1) 516 (W O2) 517 (B D3) 518 (W N6) 519 (B D5) 520 (W P9) 521 (B O7) 522 (W E4) 523 (B M6) 524 (W M7) 525 (B P7) 526 (W G19) 527 (B O6) 528 (W P3) 529 (B E2) 530 (W L4) 531 (B G2) 532 (W T6) 533 (B C3) 534 (W T7) 535 (B N5) 536 (W N4) 537 (B Q2) 538 (W T4) 539 (B T2) 540 (W J1) 541 (B S1) 542 (W T5) 543 (B S5) 544 (W B1) 545 (B Q6) 546 (W C4) 547 (B Q7) 548 (W Q4) 549 (B M2) 550 (

W R6) 551 (B R4) 552 (W S2) 553 (B L5) 554 (W G3) 555 (B P6) 556 (W B19) 557 (B O3) 558 (W N7) 559 (B R1) 560 (W M1) 561 (B F4) 562 (W H1) 563 (B S7) 564 (W S15) 565 (B S3) 566 (W A7) 567 (B P5) 568 (W C17) 569 (B P2) 570 (W pass) 571 (B P1) 572 (W O1) 573 (B Q1) 574 (W G12) 575 (B R2) 576 (W A19) 577 (B K1) 578 (W G8) 579 (B G1) 580 (W M5) 581 (B D4) 582 (W Q5) 583 (B F3) 584 (W F18) 585 (B H2) 586 (W R5) 587 (B pass) 588 (W P4) 589 (B J3) 590 (W B4) 591 (B T8) 592 (W H3) 593 (B O4) 594 (W K9) 595 (B pass) 596 (W M6) 597 (B N3) 598 (W O5) 599 (B T16) 600 (W J2) 601 (B R3) 602 (W L7) 603 (B T17) 604 (W A6) 605 (B pass) 606 (W E3) 607 (B N8) 608 (W H14) 609 (B C3) 610 (W D3) 611 (B F1) 612 (W S11) 613 (B S6) 614 (W P15) 615 (B S17) 616 (W R18) 617 (B N5) 618 (W D5) 619 (B Q16) 620 (W Q14) 621 (B S4) 622 (W K5) 623 (B Q3) 624 (W O8) 625 (B O5) 626 (W M4) 627 (B S2) 628 (W K2) 629 (B L1) 630 (W K8) 631 (B T1) 632 (W P11) 633 (B N2) 634 (W M1) 635 (B pass) 636 (W pass) Game has ended.
Score: W+233.5

**Comment by l1t1** [comment]
"resignation_percent": "0",

**Comment by grolich** [comment]
TLDR version: without some games where leela is prohibited from resigning (ever) we'll get too many wrong resignation games. There seems to be no way around this.

1) In a certain percentage of games LZ is told to play without resigning (no matter the situation). 
The whole point of "no resignation" games is to make sure the net doesn't wrongfully resign often, and removing such games will likely have a negative effect. These games are a "necessary evil" in the sense of using more computation power/time.  

2) Not sure what you mean by "decide by MC only", as the monte carlo tree search in LZ (as in AGZ) no longer does rollouts to the end of the game (ever), rather, it completely trusts the network value at the leaf nodes of the search. (that is another reason why no-resignation games are so important).

**Comment by gjm11** [comment]
There's a definite reason why some games have to run to something like completion. Not only (as grolich) says to avoid too many wrong resignations; also to keep the network's learning anchored to reality.

However, it seems to me that it could maybe be improved in a couple of related ways.

First -- and something like this has been suggested recently in the comments on another issue, I forget which one -- it's not clear that LZ is gaining much from the latest moves in the game, many of which are rather bad. It's possible that it might do better if training is done only on the earlier portion of the game. (Still using the final game _result_, though.)

Second, once the game is far enough on that normally it _would_ have been win by resignation, it's not clear that much is gained by having both sides do extensive searching on each move. Perhaps the results would be as good (especially if training ignores the last part of the game) if "late" moves were played just using NN policy output. Or, a bit more safely, using (say) 50-100 visits per move.

("Late" could mean something like "after one player would have resigned, in a normal game". Or, if the information isn't there to determine when that would have been, something like "after move 300".)

At the moment, although relatively few games are played in no-resign mode, a substantial fraction of the computer time used for self-play games is being used for no-resign games -- because they tend to be so much longer. That seems wasteful.

**Comment by jokkebk** [comment]
This was discussed at least 99 issues ago at #1509 and I also think that with 25 % resign 0 games that take 2.5x as long as normal games, almost 30 % of contributors cpu/gpu time is wasted on low quality material. Lowering visits would definitely be a great improvement.

**Comment by Ishinoshita** [comment]
@jokkebk: Resign can happen very early, as early as before 90 moves (have a look at the distribution of game length). It's important to check that such games cannot be reversed, with full strength play, i.e 3200v. However, the likelyhood a game reversal happening beyond 300 moves is very low IMO. So, based on your idea, and provided the move count is above 300 (or 360 or some level to be defined),  the game can indeed be adjudicated by a much weaker player, i..e lower visits search. Since the game is very advanced, it could as well been adjudicated by a 1 po endgame. Policy head alone should be well above 4D-5D amateur, a level very sufficient to play the game through completion and check the resign validity. The benefit of playing a percentage of no-resign would cost almost nothing (1po vs 3200 speedup) from 300 to 600 moves (~the average no-resign game length). Checking that 1po, starting from 300 moves count, produces the same result as 3200v all game long, with a high confidence level, should be quite straigtforward.

**Comment by sudosandwich** [comment]
"as the monte carlo tree search in LZ (as in AGZ) no longer does rollouts to the end of the game (ever)"

I was not aware of this

**Comment by sudosandwich** [comment]
I still believe a heuristic could work though

**Comment by Ishinoshita** [comment]
Your point is still valid. Playing games from 300 to 600+ moves count with 3200v might not be a fatality. Is roughly 15% of the total cpu/gpu workload that goes into these moves.

**Comment by sudosandwich** [comment]
My suggestion would be a cutoff at some move number. I mean 600 moves is a LOT of time

**Comment by Ishinoshita** [comment]
Cutoff is already what is done on resign game, with high computational benefit ( from 600+ moves down to ~175 moves). That's why we need to let some games unroll till end, to check the validity if the heuristic "if the value is below x%, that game can be safely resigned". Now, the question moves to what computational effort is needed to safely play a game till end, so that the outcome is reliable. Here come @jokkebk suggestion and mine.
